### PR TITLE
add new 3d tests

### PR DESCRIPTION
**Pull request by benoitdm-oslandia**
Add new 3D tests to assert the behaviours of depth buffer capture, ambient occlusion and debug map displays. This is done to avoid regressions as start point for the [FrameGraph refactoring](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/259). 

[X] can be backported

cc @ptitjano

### CODE REVIEW

File: src/gui/attributetable/qgsfeaturelistview.cpp [hunk 1]
```diff
@@ -183,6 +183,18 @@ void QgsFeatureListView::editSelectionChanged( const QItemSelection &deselected,
     viewport()->update( visualRegionForSelection( localDeselected ) | visualRegionForSelection( localSelected ) );
   }
 
+  mLastEditSelectionFid = QgsFeatureId();
```
Comment (nyalldawson) [inline src/gui/attributetable/qgsfeaturelistview.cpp:186]: can you rebase? there's a lot of unrelated commits here

Comment (benoitdm-oslandia) [reply src/gui/attributetable/qgsfeaturelistview.cpp]: done !

File: src/3d/qgscameracontroller.cpp [hunk 2]
```diff
@@ -232,9 +232,9 @@ double QgsCameraController::sampleDepthBuffer( const QImage &buffer, int px, int
   // Returns the average of depth values that are not 1 (void area)
   depth = 0;
   int samplesCount = 0;
-  for ( int x = 0; x < mDepthBufferImage.width(); ++x )
+  for ( int x = 0; x < buffer.width(); ++x )
```
Comment (ptitjano) [inline src/3d/qgscameracontroller.cpp:235]: Where does this change come from? Does this fix an issue?

File: src/3d/qgscameracontroller.cpp [hunk 3]
```diff
@@ -232,9 +232,9 @@ double QgsCameraController::sampleDepthBuffer( const QImage &buffer, int px, int
...
   {
-    for ( int y = 0; y < mDepthBufferImage.height(); ++y )
+    for ( int y = 0; y < buffer.height(); ++y )
```
Comment (ptitjano) [inline src/3d/qgscameracontroller.cpp:237]: same question.

File: tests/src/3d/testqgs3drendering.cpp [hunk 4]
```diff
@@ -106,6 +110,63 @@ class TestQgs3DRendering : public QgsTest
...
+  // compute image min/max depth values
+  double minV = 9999999.0;
+  double maxV = -9999999.0;
+  for ( int x = 0; x < grayImage.width(); x++ )
+  {
+    for ( int y = 0; y < grayImage.height(); y++ )
+    {
+      double d = Qgs3DUtils::decodeDepth( depthImage.pixel( x, y ) );
+      if ( d > maxV ) maxV = d;
+      else if ( d < minV ) minV = d;
```
Comment (ptitjano) [inline tests/src/3d/testqgs3drendering.cpp:152]: nitpick: please avoid one line `if` - `else`

File: tests/src/3d/testqgs3drendering.cpp [hunk 5]
```diff
@@ -1408,5 +1469,155 @@ void TestQgs3DRendering::testFilteredExtrudedPolygons()
...
+  QGSCOMPARENEARVECTOR3D( testCam->cameraPose()->centerPoint(), QVector3D( -32.7, 224.6, 185.5 ), 1.0 );
+  QGSCOMPARENEAR( testCam->cameraPose()->distanceFromCenterPoint(), 955.4, 1.0 );
+  QCOMPARE( testCam->cumulatedWheelY(), 0 );
+
+  // Second wheel action
+  QWheelEvent wheelEvent2( midPos, midPos, QPoint(), QPoint( 0, 120 ),
+                           Qt::NoButton, Qt::NoModifier, Qt::NoScrollPhase,
+                           false, Qt::MouseEventSynthesizedByApplication );
+  testCam->superOnWheel( new Qt3DInput::QWheelEvent( wheelEvent2 ) );
+  qDebug() << "Checking depth after onWheel called (bis)" ;
```
Comment (ptitjano) [inline tests/src/3d/testqgs3drendering.cpp:1621]: remove this line or change it to a proper message.

File: src/3d/qgscameracontroller.cpp (hunk 3 above)
Comment (benoitdm-oslandia) [reply src/3d/qgscameracontroller.cpp:237]: same response :)

File: tests/src/3d/testqgs3drendering.cpp [hunk 6]
```diff
@@ -1408,5 +1469,155 @@ void TestQgs3DRendering::testFilteredExtrudedPolygons()
...
+  QMouseEvent mouseEvent( QEvent::MouseButtonPress, QPointF( midPos.x(), midPos.y() ),
+                          Qt::MiddleButton, Qt::MiddleButton, Qt::NoModifier );
+  testCam->superOnMousePressed( new Qt3DInput::QMouseEvent( mouseEvent ) );
+
+  // First wheel action
+  QWheelEvent wheelEvent( midPos, midPos, QPoint(), QPoint( 0, 120 ),
+                          Qt::NoButton, Qt::NoModifier, Qt::NoScrollPhase,
+                          false, Qt::MouseEventSynthesizedByApplication );
+  testCam->superOnWheel( new Qt3DInput::QWheelEvent( wheelEvent ) );
+  qDebug() << "Checking depth after onWheel called" ;
```
Comment (ptitjano) [inline tests/src/3d/testqgs3drendering.cpp:1601]: remove this line or change it to a proper message.

File: tests/src/3d/testqgs3drendering.cpp [hunk 7]
```diff
@@ -1408,5 +1469,155 @@ void TestQgs3DRendering::testFilteredExtrudedPolygons()
...
+  QGSCOMPARENEARVECTOR3D( testCam->cameraPose()->centerPoint(), QVector3D( -32.5, 223.5, 184.7 ), 1.0 );
+  QGSCOMPARENEAR( testCam->cameraPose()->distanceFromCenterPoint(), 757.4, 1.0 );
+  QCOMPARE( testCam->cumulatedWheelY(), 0 );
+
+  // Third wheel action
+  QWheelEvent wheelEvent3( midPos, midPos, QPoint( ), QPoint( 0, 480 ),
+                           Qt::NoButton, Qt::NoModifier, Qt::NoScrollPhase,
+                           false, Qt::MouseEventSynthesizedByApplication );
+  testCam->superOnWheel( new Qt3DInput::QWheelEvent( wheelEvent3 ) );
+  qDebug() << "Checking depth after onWheel called (ter)" ;
```
Comment (ptitjano) [inline tests/src/3d/testqgs3drendering.cpp:1641]: remove this line or change it to a proper message.

File: tests/src/3d/testqgs3drendering.cpp [hunk 8]
```diff
@@ -1408,5 +1469,155 @@ void TestQgs3DRendering::testFilteredExtrudedPolygons()
...
+  QGSCOMPARENEARVECTOR3D( testCam->cameraPose()->centerPoint(), QVector3D( -32.4, 222.8, 184.1 ), 1.0 );
+  QGSCOMPARENEAR( testCam->cameraPose()->distanceFromCenterPoint(), 126.4, 0.1 );
+  QCOMPARE( testCam->cumulatedWheelY(), 0 );
+
+  // Fourth wheel action
+  QWheelEvent wheelEvent4( midPos, midPos, QPoint(), QPoint( 0, 120 ),
+                           Qt::NoButton, Qt::NoModifier, Qt::NoScrollPhase,
+                           false, Qt::MouseEventSynthesizedByApplication );
+  testCam->superOnWheel( new Qt3DInput::QWheelEvent( wheelEvent4 ) );
+  qDebug() << "Checking depth after onWheel called (quad)" ;
```
Comment (ptitjano) [inline tests/src/3d/testqgs3drendering.cpp:1661]: remove this line or change it to a proper message.

File: tests/src/3d/testqgs3drendering.cpp [hunk 9]
```diff
@@ -1408,5 +1469,155 @@ void TestQgs3DRendering::testFilteredExtrudedPolygons()
...
+
+  scene->cameraController()->depthBufferCaptured( depthImage );
+
+  QGSCOMPARENEARVECTOR3D( testCam->zoomPoint(), QVector3D( -32.3, 221.7, 183.2 ), 1.0 );
+  QGSCOMPARENEARVECTOR3D( testCam->cameraPose()->centerPoint(), QVector3D( -32.3, 221.7, 183.2 ), 1.0 );
+  QGSCOMPARENEAR( testCam->cameraPose()->distanceFromCenterPoint(), 101.1, 0.1 );
+  QCOMPARE( testCam->cumulatedWheelY(), 0 );
+
+  QVector3D diff = scene->cameraController()->camera()->position() - startPos;
+  qDebug() << "Checking camera position";
```
Comment (ptitjano) [inline tests/src/3d/testqgs3drendering.cpp:1677]: remove this line or change it to a proper message.

File: tests/src/3d/testqgs3drendering.cpp [hunk 10]
```diff
@@ -91,10 +93,12 @@ class TestQgs3DRendering : public QgsTest
...
     void testFilteredExtrudedPolygons();
+    void testDepthBuffer();
 
   private:
     // color tolerance < 2 was failing polygon3d_extrusion test
     bool renderCheck( const QString &testName, QImage &image, int mismatchCount = 0, int colorTolerance = 2 );
+    QImage convertDepthImageToGray16Image( const QImage &depthImage );
 
     std::unique_ptr<QgsProject> mProject;
     QgsRasterLayer *mLayerDtm;
```
Comment (ptitjano) [inline tests/src/3d/testqgs3drendering.cpp:106]: `QgsRasterLayer *mLayerDtm = nullptr;`

**Comment by ptitjano** [comment]
Please, integrate the last commit in the appropriate previous ones.

File: tests/src/3d/testqgs3drendering.cpp (hunk 4 above)
Comment (benoitdm-oslandia) [reply tests/src/3d/testqgs3drendering.cpp:152]: ok

File: tests/src/3d/testqgs3drendering.cpp (hunk 6 above)
Comment (benoitdm-oslandia) [reply tests/src/3d/testqgs3drendering.cpp:1601]: done!

File: tests/src/3d/testqgs3drendering.cpp (hunk 7 above)
Comment (benoitdm-oslandia) [reply tests/src/3d/testqgs3drendering.cpp:1641]: done!

File: tests/src/3d/testqgs3drendering.cpp (hunk 9 above)
Comment (benoitdm-oslandia) [reply tests/src/3d/testqgs3drendering.cpp:1677]: done!

File: src/3d/qgscameracontroller.cpp (hunk 2 above)
Comment (ptitjano) [reply src/3d/qgscameracontroller.cpp:235]: It fixes an issue. In that case, it should be part of a separate commit. This way, this fix could be easily backported.

Comment (benoitdm-oslandia) [reply src/3d/qgscameracontroller.cpp:235]: oki doki

**Review by lbartoletti — APPROVED** [review]

**Comment by ptitjano** [comment]
The bug fix has been merged as part of https://github.com/qgis/QGIS/pull/53474

This MR now only contains new 3d tests for: 
- `sampleDepthBuffer`
- ambient occlusion
- debug map

cc @wonder-sk @uclaros

File: tests/src/3d/testqgs3drendering.cpp [hunk 11]
```diff
@@ -1408,5 +1471,274 @@ void TestQgs3DRendering::testFilteredExtrudedPolygons()
...
+
+  // =========== set camera position
+  scene->cameraController()->setLookingAtPoint( QVector3D( 0, 0, 0 ), 400, 50, 10 );
+
+  QgsAmbientOcclusionSettings aoSettings = mapSettings.ambientOcclusionSettings();
+  aoSettings.setEnabled( false );
+  mapSettings.setAmbientOcclusionSettings( aoSettings );
+
+  QImage img = Qgs3DUtils::captureSceneImage( engine, scene );
+  QVERIFY( renderCheck( "ambient_occlusion_1", img, 40, 15 ) );
```
Comment (uclaros) [inline tests/src/3d/testqgs3drendering.cpp:1523]: I'd suggest skipping this one, as ambient occlusion is disabled, or at least run it after the `enabled` test so we can test it actually gets disabled.
You could also rename the results to `ambient_occlusion_on` / `_off`

File: tests/src/3d/testqgs3drendering.cpp [hunk 12]
```diff
@@ -1408,5 +1471,274 @@ void TestQgs3DRendering::testFilteredExtrudedPolygons()
...
+  mapSettings.setDebugDepthMapSettings( true, Qt::Corner::BottomRightCorner, 0.5 );
+
+  QImage img = Qgs3DUtils::captureSceneImage( engine, scene );
+  QVERIFY( renderCheck( "debug_map_1", img, 100, 15 ) );
+
+  // =========== activate debug shadow map
+  mapSettings.setDebugShadowMapSettings( true, Qt::Corner::TopLeftCorner, 0.5 );
+
+  img = Qgs3DUtils::captureSceneImage( engine, scene );
+  QVERIFY( renderCheck( "debug_map_2", img, 100, 15 ) );
```
Comment (uclaros) [inline tests/src/3d/testqgs3drendering.cpp:1735]: This doesn't seem to actually test anything: `expected_debug_map_1.png` and `expected_debug_map_2.png` are binary equal.

If you make proper rendering checks for the _debugShadowMap_ and _debugDepthMap_ windows you could also eliminate the need for `convertDepthImageToGray16Image()` and replace it with one check against a map with `mapSettings.setDebugDepthMapSettings( true, Qt::Corner::BottomRightCorner, **1** );`

File: tests/src/3d/testqgs3drendering.cpp [hunk 13]
```diff
@@ -1408,5 +1471,274 @@ void TestQgs3DRendering::testFilteredExtrudedPolygons()
...
+
+  img = Qgs3DUtils::captureSceneImage( engine, scene );
+  QVERIFY( renderCheck( "ambient_occlusion_2", img, 40, 15 ) );
+
+  delete scene;
+  mapSettings.setLayers( {} );
+  demTerrain->deleteLater();
+}
+
+void TestQgs3DRendering::testDepthBuffer()
```
Comment (uclaros) [inline tests/src/3d/testqgs3drendering.cpp:1536]: This test seems to test both mouse wheel actions and depth map generation. Could you maybe split that into two distinct tests? Maybe _rendering_ is not the best place for the mouse wheel test anyway.

**Review by uclaros — COMMENTED** [review]
Sorry for being late, but I'm afraid this was not adequately reviewed before merging...

File: tests/src/3d/testqgs3drendering.cpp [hunk 14]
```diff
@@ -91,21 +93,82 @@ class TestQgs3DRendering : public QgsTest
...
+  // transform depth value to gray value
+  double factor = 65635.0 / ( maxV - minV );
+  for ( int x = 0; x < grayImage.width(); x++ )
+  {
+    for ( int y = 0; y < grayImage.height(); y++ )
+    {
+      double d = Qgs3DUtils::decodeDepth( depthImage.pixel( x, y ) );
+      unsigned short v = ( unsigned short )( factor * ( d - minV ) );
+      QRgba64 col = QRgba64::fromRgba64( v, v, v, ( quint16 )65635 );
+      grayImage.setPixelColor( x, y, QColor( col ) );
```
[the remark this answers is not in the document]
Comment (benoitdm-oslandia) [reply tests/src/3d/testqgs3drendering.cpp:165]: fixing

File: tests/src/3d/testqgs3drendering.cpp [hunk 15]
```diff
@@ -91,21 +93,82 @@ class TestQgs3DRendering : public QgsTest
...
+  }
+
+  // transform depth value to gray value
+  double factor = 65635.0 / ( maxV - minV );
+  for ( int x = 0; x < grayImage.width(); x++ )
+  {
+    for ( int y = 0; y < grayImage.height(); y++ )
+    {
+      double d = Qgs3DUtils::decodeDepth( depthImage.pixel( x, y ) );
+      unsigned short v = ( unsigned short )( factor * ( d - minV ) );
```
[the remark this answers is not in the document]
Comment (benoitdm-oslandia) [reply tests/src/3d/testqgs3drendering.cpp:163]: checked

File: tests/src/3d/testqgs3drendering.cpp (hunk 11 above)
Comment (benoitdm-oslandia) [reply tests/src/3d/testqgs3drendering.cpp:1523]: fixed

File: tests/src/3d/testqgs3drendering.cpp (hunk 12 above)
Comment (benoitdm-oslandia) [reply tests/src/3d/testqgs3drendering.cpp:1735]: In fact the test fails because the 2 images are the same. It was my mistake to copy the expected image from another test to have a default image to check against but in fact the `captureSceneImage` failed to render the debug maps

File: tests/src/3d/testqgs3drendering.cpp (hunk 13 above)
Comment (benoitdm-oslandia) [reply tests/src/3d/testqgs3drendering.cpp:1536]: The main goal is to check if the depth buffer evolves according to the camera moves and the depth buffer generation is called from wheel action.

File: tests/src/3d/testqgs3drendering.cpp (hunk 12 above)
Comment (benoitdm-oslandia) [reply tests/src/3d/testqgs3drendering.cpp:1735]: Hi @uclaros, I have been looking at this issue all day but I can not find why the debug texture map is not working at all. Any clues?

Comment (uclaros) [reply tests/src/3d/testqgs3drendering.cpp:1735]: Hmm, I don't know, I would need to investigate. It could be related to it being on a different viewport, so I would start by checking that the 3dAxis (which is also on a different viewport) is being exported.
